instrumentation/opentelemetry-instrumentation-pika/src/opentelemetry/instrumentation/pika/pika_instrumentor.py

```diff
@@ -0,0 +1,105 @@
+# limitations under the License.
+from logging import getLogger
+from typing import Any, Callable, Collection, Dict, Optional
+
+from pika.channel import Channel
+
+from opentelemetry import trace
+from opentelemetry.instrumentation.instrumentor import BaseInstrumentor
+from opentelemetry.instrumentation.pika import utils
+from opentelemetry.instrumentation.pika.package import _instruments
+from opentelemetry.instrumentation.pika.version import __version__
+from opentelemetry.trace import Tracer, TracerProvider
+
+_LOG = getLogger(__name__)
+CTX_KEY = "__otel_task_span"
+
+FUNCTIONS_TO_UNINSTRUMENT = ["basic_publish"]
+
+
+class PikaInstrumentor(BaseInstrumentor):  # type: ignore
+    @staticmethod
+    def _instrument_consumers(
+        consumers_dict: Dict[str, Callable[..., Any]], tracer: Tracer
+    ) -> Any:
+        for key, callback in consumers_dict.items():
+            decorated_callback = utils.decorate_callback(callback, tracer, key)
+            setattr(decorated_callback, "_original_callback", callback)
+            consumers_dict[key] = decorated_callback
+
+    @staticmethod
+    def _instrument_basic_publish(channel: Channel, tracer: Tracer) -> None:
+        original_function = getattr(channel, "basic_publish")
+        decorated_function = utils.decorate_basic_publish(
+            original_function, channel, tracer
+        )
+        setattr(decorated_function, "_original_function", original_function)
+        channel.__setattr__("basic_publish", decorated_function)
+        channel.basic_publish = decorated_function
+
+    @staticmethod
+    def _instrument_channel_functions(
+        channel: Channel, tracer: Tracer
+    ) -> None:
+        if hasattr(channel, "basic_publish"):
+            PikaInstrumentor._instrument_basic_publish(channel, tracer)
+
+    @staticmethod
+    def _uninstrument_channel_functions(channel: Channel) -> None:
+        for function_name in FUNCTIONS_TO_UNINSTRUMENT:
+            if not hasattr(channel, function_name):
+                continue
+            function = getattr(channel, function_name)
+            if hasattr(function, "_original_function"):
+                channel.__setattr__(function_name, function._original_function)
+
+    @staticmethod
+    def instrument_channel(
+        channel: Channel, tracer_provider: Optional[TracerProvider] = None,
+    ) -> None:
+        if not hasattr(channel, "_impl"):
+            _LOG.error("Could not find implementation for provided channel!")
+            return
+        tracer = trace.get_tracer(__name__, __version__, tracer_provider)
+        channel.__setattr__("__opentelemetry_tracer", tracer)
+        if channel._impl._consumers:
+            PikaInstrumentor._instrument_consumers(
+                channel._impl._consumers, tracer
+            )
+        PikaInstrumentor._instrument_channel_functions(channel, tracer)
+
+    def _instrument(self, **kwargs: Dict[str, Any]) -> None:
```

Comment:
This doesn't look right. `_instrument` must instrument a package globally, not a single object provided by a package. This is what makes it possible to auto instrument things with the `opentelemetry-instrument` command.

Reply:
Fixed